### PR TITLE
TCP Full Duplex Server Client Communication

### PR DESCRIPTION
**Pull request by NVombat**
#### Description of Change

Implemented a Full Duplex Communication server client model using a TCP socket

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C/pull/856"></a>

### CODE REVIEW

File: client_server/tcp_full_duplex_server.c [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+#include <arpa/inet.h>  /// For the type in_addr_t and in_port_t
+#include <netdb.h>  /// For structures returned by the network database library - formatted internet addresses and port numbers
+#include <netinet/in.h>  /// For in_addr and sockaddr_in structures
+#include <stdint.h>      /// For specific bit size values of variables
+#include <stdio.h>  /// Variable types, several macros, and various functions for performing input and output
+#include <stdlib.h>  /// Variable types, several macros, and various functions for performing general functions
+#include <string.h>  /// Various functions for manipulating arrays of characters
+#include <sys/socket.h>  /// For macro definitions related to the creation of sockets
+#include <sys/types.h>  /// For definitions to allow for the porting of BSD programs
+#include <unistd.h>  /// For miscellaneous symbolic constants and types, and miscellaneous functions
```
Comment (Panquesito7) [inline client_server/tcp_full_duplex_server.c:29]:
```suggestion
#include <arpa/inet.h>             /// For the type in_addr_t and in_port_t
#include <netdb.h>                /// For structures returned by the network database library - formatted internet addresses and port numbers
#include <netinet/in.h>          /// For in_addr and sockaddr_in structures
#include <stdint.h>             /// For specific bit size values of variables
#include <stdio.h>             /// Variable types, several macros, and various functions for performing input and output
#include <stdlib.h>           /// Variable types, several macros, and various functions for performing general functions
#include <string.h>          /// Various functions for manipulating arrays of characters
#include <sys/socket.h>     /// For macro definitions related to the creation of sockets
#include <sys/types.h>     /// For definitions to allow for the porting of BSD programs
#include <unistd.h>       /// For miscellaneous symbolic constants and types, and miscellaneous functions
```

File: client_server/tcp_full_duplex_client.c [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+#include <arpa/inet.h>  /// For the type in_addr_t and in_port_t
+#include <netdb.h>  /// For structures returned by the network database library - formatted internet addresses and port numbers
+#include <netinet/in.h>  /// For in_addr and sockaddr_in structures
+#include <stdint.h>      /// For specific bit size values of variables
+#include <stdio.h>  /// Variable types, several macros, and various functions for performing input and output
+#include <stdlib.h>  /// Variable types, several macros, and various functions for performing general functions
+#include <string.h>  /// Various functions for manipulating arrays of characters
+#include <sys/socket.h>  /// For macro definitions related to the creation of sockets
+#include <sys/types.h>  /// For definitions to allow for the porting of BSD programs
+#include <unistd.h>  /// For miscellaneous symbolic constants and types, and miscellaneous functions
```
Comment (Panquesito7) [inline client_server/tcp_full_duplex_client.c:29]: Please do the same here as above.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Awesome work! 😄👍
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/NVombat/C/actions

File: client_server/tcp_full_duplex_client.c (hunk 2 above)
Comment (NVombat) [reply client_server/tcp_full_duplex_client.c]: I am unable to do this from my IDE due to the default formatting settings - I'll try changing this on Github itself

File: client_server/tcp_full_duplex_client.c [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+ * It prints `str` and an implementation-defined error
+ * message corresponding to the global variable `errno`.
+ * @returns void
+ */
+void error()
+{
+    perror("Socket Creation Failed");
+    exit(EXIT_FAILURE);
+}
+
```
Comment (Panquesito7) [inline client_server/tcp_full_duplex_client.c:45]: _If applicable_: add self-test implementations using `assert`.

File: client_server/tcp_full_duplex_server.c [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+ * It prints `str` and an implementation-defined error
+ * message corresponding to the global variable `errno`.
+ * @returns void
+ */
+void error()
+{
+    perror("Socket Creation Failed");
+    exit(EXIT_FAILURE);
+}
+
```
Comment (Panquesito7) [inline client_server/tcp_full_duplex_server.c:44]: _If applicable_: add self-test implementations using `assert`.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: client_server/tcp_full_duplex_client.c (hunk 3 above)
Comment (NVombat) [reply client_server/tcp_full_duplex_client.c:44]: It wont be applicable in this case as the only user input/variable data is the message being sent which can be any format or string anyway

File: client_server/tcp_full_duplex_server.c (hunk 4 above)
Comment (NVombat) [reply client_server/tcp_full_duplex_server.c:44]: It wont be applicable in this case as the only user input/variable data is the message being sent which can be any format or string anyway

**Review by amino19 — APPROVED** [review]

**Comment by Panquesito7** [comment]
Thanks for reviewing, @amino19! 😄

**Pull request merged**